compiler/src/org.graalvm.compiler.asm.amd64/src/org/graalvm/compiler/asm/amd64/AMD64MacroAssembler.java

```diff
@@ -773,4 +777,412 @@ public void stringIndexOf(Register str1, Register str2,
+            jcc(ConditionFlag.NotZero, L_sse_loop);
+
+            Label L_copy_chars = new Label();
+            // Test and compress another 8 chars before final tail copy.
+            bind(L_sse_tail);
+            movl(res, len);
+            andl(res, -8);
+            jccb(ConditionFlag.Zero, L_copy_chars);
+            andl(len, 8 - 1);
+
+            movdl(vtmp1, tmp);
+            pshufd(vtmp1, vtmp1, 0);    // Store Unicode mask in 'vtmp1'.
+            pxor(vtmp3, vtmp3);
+
+            movdqu(vtmp2, new AMD64Address(src));
+            ptest(vtmp2, vtmp1);        // Check for Unicode chars in vector.
+            jccb(ConditionFlag.NotZero, L_return_zero);
+            packuswb(vtmp2, vtmp3);     // Only ASCII chars; compress each to a byte.
+            movq(new AMD64Address(dst), vtmp2);
+            addq(src, 16);
+            addq(dst, 8);
+
+            bind(L_copy_chars);
+        }
+
+        // Compress any remaining characters using a vanilla implementation.
+        testl(len, len);
+        jccb(ConditionFlag.Zero, L_return_length);
+        leaq(src, new AMD64Address(src, len, Scale.Times2));
+        leaq(dst, new AMD64Address(dst, len, Scale.Times1));
+        negq(len);
+
+        Label L_copy_chars_loop = new Label();
+        // Compress a single character per iteration.
+        bind(L_copy_chars_loop);
+        movzwl(res, new AMD64Address(src, len, Scale.Times2));
+        testl(res, 0xff00);     // Check if Unicode character.
+        jccb(ConditionFlag.NotZero, L_return_zero);
+        // An ASCII character; compress to a byte.
+        movb(new AMD64Address(dst, len, Scale.Times1), res);
+        incrementq(len, 1);
+        jcc(ConditionFlag.NotZero, L_copy_chars_loop);
+
+        // If compression succeeded, return the length.
+        bind(L_return_length);
+        pop(res);
+        jmpb(L_done);
+
+        // If compression failed, return 0.
+        bind(L_return_zero);
+        xorl(res, res);
+        addq(rsp, 8 /*wordSize*/);
+
+        bind(L_done);
+    }
+
+    /* Copy (inflate) a Latin1 string using a byte[] array representation into
+     * a UTF16 string using a char[] array representation.
+     *
+     *   src    (rsi) the start address of source byte[] to be inflated
+     *   dst    (rdi) the start address of destination char[] array
+     *   len    (rdx) the length
+     *   vtmp   (xmm)
+     *   tmp    (gpr)
+     */
+    public void byte_array_inflate(Register src, Register dst, Register len,
+                                   Register vtmp, Register tmp) {
+
+        assert vtmp.getRegisterCategory().equals(AMD64.XMM);
+
+        Label L_done = new Label();
+        Label L_below_threshold = new Label();
+
+        assert src.number != dst.number && src.number != len.number && src.number != tmp.number;
+        assert dst.number != len.number && dst.number != tmp.number;
+        assert len.number != tmp.number;
+
+        if (UseAVX > 2 && (supports(CPUFeature.AVX512BW) &&
```

Comment:
I don't see that "config.useAVX" (from HS: "UseAVX") is actually propagated to the feature setup (not in the code I have). The "config.useSSE" is indeed but the interpretation has been changed such that it is no longer possible to turn SSE code generation off with "UseSSE=0". Moreover, the Graal interpretation of "config.useSSE" has been made AMD64 exclusive which is not the case in HS. In short, I just bailed out to take an orthogonal approach for now.

Reply:
Well we don't expose UseAVX as value since everything is expressed in terms of CPUFeatures.  If there are levels above AVX2 that we need to support then they need to be added to CPUFeatures in JVMCI and propagated that way.  I don't think our assembly has working support for EVEX as the point either.

Reply:
The path by which UseAVX propagates into CPUFeatures is indeed complex but it goes through VM_Version::Feature_Flags which a driven by UseAVX.  For levels above 2 I think `jdk.vm.ci.hotspot.amd64.AMD64HotSpotJVMCIBackendFactory#computeFeatures` will have to directly test against UseAVX since there are no FeatureFlags for them.